Testing: Remove enzyme completely

tyxla opened the pull request:
## What?
This PR gets rid of good old `enzyme` now that we've migrated all tests to use `@testing-library/react`.

Fixes #17249.

## Why?
Enzyme has been a major blocker for upgrading to React 18, so it's great to see it gone!

## How?
We're getting rid of the enzyme dependency, its React 17 adapter and snapshot serializers.

~This PR includes #44492 and #44493 and should be rebased after they land.~

We're also consequently getting rid of the last usage of `underscore`, as explained [here](https://github.com/WordPress/gutenberg/pull/44494#issuecomment-1259449308).

## Testing Instructions
Verify all checks are green.

## Notes

This change definitely deserves a dev note.

tyxla committed:
Remove enzyme completely

tyxla committed:
Add CHANGELOG entry

gziolo commented on packages/jest-preset-default/CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
+### Breaking Change
+
+-   Testing: Remove enzyme completely ([#44494](https://github.com/WordPress/gutenberg/pull/44494)).
```
Let's leave a note in the README how to bring back `enzyme` in case someone is using it.

tyxla replied:
Good call. Done in https://github.com/WordPress/gutenberg/pull/44494/commits/8e5c8c2fa402fa7b9195aa8e9df72c24e9e9e732. Feel free to contribute if you have any additional ideas on what to add there.

gziolo replied:
Excellent, can we also link it from the CHANGELOG so it's easier to discover.

tyxla replied:
Good point, addressed in bde4b519f151d0395482ac67dc272d9be3a25b95

gziolo reviewed (commented):
Excellent work on replacing Enzyme with React Testing Library.

All code changes look great. We should ensure that we have documented the case when some project use `enzyme` so they know what to install and configure so they can continue using their current tests.

There is one open question for future work. Should we make RTL a part of the default Jest preset we ship with `@wordpress/scripts`?

tyxla committed:
Document how to use enzyme

Mamaduka commented:
Fantastic work, @tyxla!

tyxla committed:
Link from CHANGELOG to README

tyxla committed:
Clarify relationship between RTL and React 18 as rationale for removing enzyme

tyxla commented:
> There is one open question for future work. Should we make RTL a part of the default Jest preset we ship with @wordpress/scripts?

We could, potentially. What benefit would that yield, though on top of always including `jest-dom` and the snapshot diffing utils?

tyxla commented:
I've rebased after merging the dependency PRs, and I think all feedback has been addressed and the PR is ready for another look 👀 .

gziolo reviewed (approved):
❤️

tyxla committed:
Remove underscore usage completely

tyxla commented:
It seems like we were implicitly relying on `underscore` to be available, which was fine so far because it's a dependency of `enzyme`. However, now that we removed `enzyme`, `underscore` was no longer provided, so the native tests started failing. The usage was straightforward - just a plain shallow `_.clone()`, which I've refactored to use object spread in https://github.com/WordPress/gutenberg/pull/44494/commits/11abc1911ef666ccc71ecb8254921e277b164f9d. That allows us to get rid of the last `underscore` usage in our codebase, which is really nice little win ✅ .

kevin940726 reviewed (approved):
LGTM 👍 Awesome!

youknowriad commented:
Nice :)

tyxla commented:
Here's a **draft dev note**, feel free to edit at will, everyone:

---

## Dropping Enzyme in favor of `@testing-library/react`

In WordPress 6.1, `enzyme` has been removed. 

Prior to this release, it was provided by the `@wordpress/jest-preset-default` package as part of the testing infrastructure. Historically, the `@wordpress/jest-preset-default` package used to use `enzyme`, but support was dropped in favor of `@testing-library/react`, the primary reason being the fact that it was incompatible with React 18, and we are aiming at unblocking the upgrade to React 18. 

If you wish to use `enzyme` anyway, you can still use it by manually providing the React 17 adapter, by following the steps below.

To install the enzyme dependency, run:

```bash
npm install --save enzyme
```

To install the React 17 adapter dependency, run:

```bash
npm install --save @wojtekmaj/enzyme-adapter-react-17
```

To use the React 17 adapter, use this in your [`setupFilesAfterEnv`](https://jestjs.io/docs/configuration#setupfilesafterenv-array) configuration:

```javascript
// It "mocks" enzyme, so that we can delay loading of
// the utility functions until enzyme is imported in tests.
// Props to @gdborton for sharing this technique in his article:
// https://medium.com/airbnb-engineering/unlocking-test-performance-migrating-from-mocha-to-jest-2796c508ec50.
let mockEnzymeSetup = false;
jest.mock( 'enzyme', () => {
	const actualEnzyme = jest.requireActual( 'enzyme' );
	if ( ! mockEnzymeSetup ) {
		mockEnzymeSetup = true;
		// Configure enzyme 3 for React, from docs: http://airbnb.io/enzyme/docs/installation/index.html
		const Adapter = jest.requireActual(
			'@wojtekmaj/enzyme-adapter-react-17'
		);
		actualEnzyme.configure( { adapter: new Adapter() } );
	}
	return actualEnzyme;
} );
```

If you also use snapshot tests with `enzyme`, you might want to add support for serializing them, through the `enzyme-to-json` package.

To install the dependency, run:

```bash
npm install --save enzyme-to-json
```

Finally, you should add `enzyme-to-json/serializer` to the array of [`snapshotSerializers`](https://jestjs.io/docs/configuration#snapshotserializers-arraystring):

```javascript
{
	snapshotSerializers: [ 'enzyme-to-json/serializer' ]
}
```

ljharb commented:
fwiw enzyme absolutely will be compatible with react 18 at some point.

tyxla commented:
> fwiw enzyme absolutely will be compatible with react 18 at some point.

Good to know, thanks @ljharb! I've edited the dev note to account for this.

ljharb commented:
I’ll also point out that the effort y’all expended on this transition could have probably helped me get enzyme compatible with react 17/18 twice over :-/

tyxla commented:
> I’ll also point out that the effort y’all expended on this transition could have probably helped me get enzyme compatible with react 17/18 twice over :-/

It's worth mentioning that this hasn't been an intentional crusade against Enzyme. Historically, Enzyme has been an indispensable utility in our testing toolchain, and we've been infinitely thankful to have it. Given that we had decided to go ahead with using `enzyme`, we would have been happy to help with the React 17/18 compatibility. 

However, recently, we've found a lot of value in using `@testing-library` for testing components. On the one hand, we find valuable the fact that we get to test the components and the way that the user interacts with them and experiences them. That makes the tests very explicit and useful to our future selves and allows us to avoid testing implementation details. With such tests, it's really easy for someone unfamiliar to understand what a component does, and how it should behave. Furthermore, using `@testing-library` has greatly helped improve the accessibility of our components and the accessibility knowledge of everyone working on this migration. 

Eventually, we could have a use for `enzyme` once more in the future, but for now, we just found other options that worked better for the project.

On a personal note, I'm curious if any progress has been made on the migration to offer first-class support for React 17 and 18. What kind of help would you be looking for to get it to the finish line?

ljharb commented:
Understood. I'm not aware of much that RTL has that enzyme's `mount` doesn't (custom hook testing is one big gap), but either way, you're obviously entitled to migrate to whatever you like :-) No shame or resentment here directed at WP or anyone else!

I just wanted to make it clear for future readers that enzyme's lack of movement is almost exclusively due to about 3-4 years worth of time where virtually no money, and only a few humans' time, have been offered. 

To get React 17 support over the finish line (which is an obvious prereq for 18 support), I'd need nonzero people to devote the time to ramp up on the project and its testing conventions, and pair with me on getting the 17 adapter passing tests and released - after which we could build the 18 adapter off of that, and evaluate how much work *that* would take.

To avoid further noise for gutenberg, if anyone's interested in discussing further, please file an issue on enzyme, or reach out to me on any of the numerous mediums I have an account, and we can chat!
